src/halfutils.c

```diff
@@ -116,6 +153,30 @@ HalfvecInnerProductF16c(int dim, half * ax, half * bx)
 
 	return distance;
 }
+
+#ifdef HAVE_AVX512FP16
+TARGET_AVX512FP16 static float
+HalfvecInnerProductAvx512Fp16(int dim, half * ax, half * bx)
+{
+	float		distance;
+	int			i;
+	int			count = (dim / 32) * 32;
+	__m512h		dist = _mm512_setzero_ph();
+
+	for (i = 0; i < count; i += 32) {
+		__m512h axi = _mm512_loadu_ph(ax+i);
+		__m512h bxi = _mm512_loadu_ph(bx+i);
+		dist = _mm512_fmadd_ph(axi, bxi, dist);
+	}
+
+	distance = (float)_mm512_reduce_add_ph(dist);
+
+	for (; i < dim; i++)
+		distance += HalfToFloat4(ax[i]) * HalfToFloat4(bx[i]);
```

Comment:
Is this auto-vectorized?  (Same question for `HalfvecL2SquaredDistanceAvx512Fp16()`.)

Reply:
I checked L2SquaredDistance and InnerProduct, and it is using AVX scalar instructions, at least with gcc-12. We'll try masked vector instructions to handle the loop remainder.

Reply:
The latest update includes masked vector instructions for the loop remainder.